log/rotate.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2022-2024 Sauce Labs Inc., all rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package log
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"sync/atomic"
+	"syscall"
+	"time"
+)
+
+func NewRotatableFile(f *os.File) *RotatableFile {
+	w := &RotatableFile{
+		ch: make(chan os.Signal, 1),
+	}
+	w.f.Store(f)
+	go w.reopenOnSIGHUP()
+	return w
+}
+
+type RotatableFile struct {
+	f  atomic.Pointer[os.File]
+	ch chan os.Signal
+}
+
+func (w *RotatableFile) file() *os.File {
+	return w.f.Load()
+}
+
+func (w *RotatableFile) Write(p []byte) (n int, err error) {
+	return w.file().Write(p)
+}
+
+var closeSignal = syscall.Signal(-1)
+
+func (w *RotatableFile) Close() error {
+	select {
+	case w.ch <- closeSignal:
+	default:
+	}
```

Comment:
I understand, it is to protect consecutive Close calls, but what If the channel is occupied with SIGHUPs and you only call close once. It's a rare case, but not impossible.

Reply:
This protects against blocking on Close() when Close is called multiple times after to Goroutine exited.

Note there is a race condition between Close and Reopen (and write) using atomics.
We could solve it by locking but then all logging would go under RW lock read path.
Given the use case I do not think it's a right approach.

There is not much we can do here tbh given this design - that's why there was no cancelation in the 1st place - it's a virtual problem.

Some of the common aids do not work here
* Closing the channel causes panics in signal and Subsequent calls to Close
* Using separate channel has the same problem
* Increasing channel buffer size does not work as you can have multiple signals queued in theory

Maybe you can suggest something that I'm missing.

Reply:
How about
1. The Close would need to call signal.Stop on the channel, so that It can safely close it. Unfortunately, it tears apart Notify from Stop.
2. use sync.Once to send closeSignal